### ISSUE TITLE
Conversation API

### ISSUE THREAD

**Issue by gabizou** [SpongePowered/SpongeAPI#432]
Bukkit had a decent API called "Conversations" where a [`Conversable`](https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/conversations/Conversable.java) would be able to converse with a plugin and vice versa. 

The core of the [`Conversation`](https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/conversations/Conversation.java) was being able to use a `Conversation` to maintain a quasi transient [`ConversationContext`](https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/conversations/ConversationContext.java) that basically was a `Map<String, Object>` to contain information pertinent to the conversation in the context between the player and the plugin.

The API had quite a bit to use, and there are some plugins that have been able to use them more robustly for interactive chat-based menus, and to be honest, I'd like to see some form of this API in SpongeAPI. 

Ideally, as the [`getConversationData(Object)`](https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/conversations/ConversationContext.java#L56) is rather abstract, I'd prefer to see some use of [`DataQuery`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/service/persistence/data/DataQuery.java) as keys. Of course, all of this is up for debate and discussion.

**Comment by ryantheleach** [comment]
Did bukkit suppress chat and join messages while the conversation was displayed?

If so, should they be queued and released to the player when the conversation releases, or just discarded?

**Comment by gabizou** [comment]
> Did bukkit suppress chat and join messages while the conversation was displayed?

It depended on the `Conversation` itself. The [`ConversationFactory`](https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/conversations/ConversationFactory.java#L52) sets up the conversation whether the chat is passed through to the rest of the server, or whether the chat is "caught" by the conversation only. As for the other bit of chat from the server, I could imagine a "bouncer" like playback created when necessary.

**Comment by Sannholm** [comment]
Can't people implement this quite easily on their own?

**Comment by ryantheleach** [comment]
Yes, but theres some advantage to having consistent look and feel between plugin UI.

**Comment by ryantheleach** [comment]
Could this be achieved by setting the players MessageSink and redirecting their messages to a plugin?
https://github.com/SpongePowered/SpongeAPI/blob/411037993587ae6f46a875198efe13c766b745a0/src/main/java/org/spongepowered/api/util/command/CommandSource.java#L74

The only thing that seems strange is the requirement of a CommandSource to be the recipient of a message... Or would creating a plugin defined CommandSource be appropriate for this use..?

**Comment by gabizou** [comment]
@ryantheleach It is achievable that way, but bucket had a nicer system where players didn't receive any chat while in the "conversation". @zml2008 would this be feasible with Text API and Messaging API?

**Comment by gabizou** [comment]
@zml2008 is this still within the realm of possibility for 3.0?

**Comment by 20zinnm** [comment]
Would this be better as a service, much like the Pagination service?

**Comment by Favorlock** [comment]
I am certainly interested in seeing a conversation API come to fruition. The Enjin Minecraft Plugin utilizes Bukkit's Conversation API, so we definitely see potential for this feature in Sponge.

**Comment by gabizou** [comment]
@20zinnm It'd likely not be a service, as there's a lot involved with `MessageChannel`s and such that require the client not to be alerted, but again, as most people have said, this is already feasible to implement directly in a plugin by itself. Perhaps @windy1 or @kashike can give light to seeing this possibly added in the future?

**Comment by kashike** [comment]
I'll look into this.

**Comment by JBYoshi** [comment]
For reference: #1532
